src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
+     */
+    public V getFirst(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return null;
         }
+        return list.get(0);
     }
 
     public boolean containsKey(K key) {
-		synchronized (table) {
-			return table.containsKey(key);
-		}
+        return this.table.containsKey(key);
     }
 
     public boolean containsElement(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v != null) && v.contains(value);
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return false;
         }
+        return list.contains(value);
     }
 
     /**
-     * Users will have to handle synchronizing.
+     * Returns mapped value collection as {@link Enumeration}
+     *
+     * <b>Note:</b> this method is deprecated
+     * please use other {@link #getAllAsList(Object)} method variant, which provides a {@link List}
+     *
+     * @param key key mapping
      */
+    @Deprecated
     public Enumeration<V> getAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-        return (v == null ?
-                new EmptyEnumeration<V>() :
-                v.elements());
+        List<V> elements = getAllAsList(key);
+        if (elements.isEmpty()) {
+            return Collections.emptyEnumeration();
+        } else {
+            return Collections.enumeration(elements);
+        }
+    }
+
+    public List<V> getAllAsList(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return Collections.unmodifiableList(list);
     }
 
     /**
      * To be used in for(x : y).
      */
     public Iterable<V> iterateAll(K key) {
-		synchronized (table) {
-			return(table.get(key));
-		}
+        return getAllAsList(key);
     }
 
     public int countAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-    	if(v != null)
-        	return v.size();
-        else
-        	return 0;
+        return getAllAsList(key).size();
     }
 
+    /**
+     * Returns mapped value collection as raw {@link Object}
+     *
+     * <b>Note:</b> this method is deprecated, this was used in previous {@code synchronized} implementation variant,
+     * please use other {@link #getAllAsList(Object)})} method variant, which provides a typed {@link List} collection.
+     *
+     * @param key key mapping
+     */
+    @Deprecated
     public Object getSync(K key) {
-		synchronized (table) {
-			return table.get(key);
-		}
+        return getAllAsList(key);
     }
 
+    /**
+     * Returns mapped value collection as array
+     *
+     * <b>Note:</b> this method is deprecated,
+     * please use other {@link #getAllAsList(Object)} method variant, which provides a typed {@link List} collection.
```

Comment:
The `@deprecated` Javadoc tag is intended for this kind of deprecation notices, and provides better IDE integration. Suggestion:
```suggestion
     * @deprecated use {@link #getAllAsList(Object)} with {@link List#toArray()} to obtain the values as object array
```